File: Euclid.Storage/Database/Access/UserDao.cs
using Euclid.Storage.Database.Data;
using NHibernate.Linq;
using System.Linq;

namespace Euclid.Storage.Database.Access
{
    public class UserDao
    {
        /// <summary>
        /// Login user with SSO ticket
        /// </summary>
        public static PlayerData Login(string username, string password)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                PlayerData playerDataAlias = null;

                var row = session.QueryOver(() => playerDataAlias)
                    .Where(() =>
                        (playerDataAlias.Name == username) &&
                        (playerDataAlias.Password == password))
                    .Take(1)
                .SingleOrDefault();

                if (row != null)
                    return row;
            }

            return null;
        }

        /// <summary>
        /// Save or update player data
        /// </summary>
        /// <param name="playerData">the player data to save</param>
        public static void SaveOrUpdate(PlayerData playerData)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                using (var transaction = session.BeginTransaction())
                {
                    try
                    {
                        session.SaveOrUpdate(playerData);
                        transaction.Commit();
                        session.Refresh(playerData);
                    }
                    catch
                    {
                        transaction.Rollback();
                    }
                }
            }
        }

        /// <summary>
        /// Get player by username
        /// </summary>
        public static PlayerData GetByName(string name)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<PlayerData>().Where(x => x.Name == name).SingleOrDefault();
            }
        }

        /// <summary>
        /// Get player by id
        /// </summary>
        public static PlayerData GetById(int id)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<PlayerData>().Where(x => x.Id == id).SingleOrDefault();
            }
        }

        /// <summary>
        /// Get player name by id
        /// </summary>
        public static string GetNameById(int id)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<PlayerData>().Select(x => x.Name).Where(x => x.Id == id).SingleOrDefault<string>();
            }
        }

        /// <summary>
        /// Get player id by name
        /// </summary>
        public static int GetIdByName(string name)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                return session.QueryOver<PlayerData>().Select(x => x.Id).Where(x => x.Name == name).SingleOrDefault<int>();
            }
        }

        /// <summary>
        /// Update last online for the last online
        /// </summary>
        public static void SaveLastOnline(PlayerData playerData)
        {
            using (var session = SessionFactoryBuilder.Instance.SessionFactory.OpenSession())
            {
                session.Query<PlayerData>().Where(x => x.Id == playerData.Id).Update(x => new PlayerData { LastOnline = playerData.LastOnline });
            }
        }
    }
}

File: euclid-storage/src/main/java/euclid/storage/database/access/UserDao.java
package euclid.storage.database.access;

import euclid.storage.database.SessionFactoryBuilder;
import euclid.storage.database.data.PlayerData;
import org.hibernate.Session;
import org.hibernate.query.MutationQuery;
import org.hibernate.query.Query;

import java.time.LocalDateTime;

public class UserDao {

    public static PlayerData login(String username, String password) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<PlayerData> query = session.createQuery(
                    "FROM PlayerData p WHERE p.name = :name AND p.password = :password", PlayerData.class);
            query.setParameter("name", username);
            query.setParameter("password", password);
            query.setMaxResults(1);
            var results = query.getResultList();
            return results.isEmpty() ? null : results.get(0);
        }
    }

    public static void saveOrUpdate(PlayerData playerData) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                session.saveOrUpdate(playerData);
                transaction.commit();
                session.refresh(playerData);
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }

    public static PlayerData getByName(String name) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<PlayerData> query = session.createQuery(
                    "FROM PlayerData p WHERE p.name = :name", PlayerData.class);
            query.setParameter("name", name);
            var results = query.getResultList();
            return results.isEmpty() ? null : results.get(0);
        }
    }

    public static PlayerData getById(int id) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            return session.find(PlayerData.class, id);
        }
    }

    public static String getNameById(int id) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<String> query = session.createQuery(
                    "SELECT p.name FROM PlayerData p WHERE p.id = :id", String.class);
            query.setParameter("id", id);
            var results = query.getResultList();
            return results.isEmpty() ? null : results.get(0);
        }
    }

    public static int getIdByName(String name) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            Query<Integer> query = session.createQuery(
                    "SELECT p.id FROM PlayerData p WHERE p.name = :name", Integer.class);
            query.setParameter("name", name);
            var results = query.getResultList();
            return results.isEmpty() ? 0 : results.get(0);
        }
    }

    public static void saveLastOnline(PlayerData playerData) {
        try (Session session = SessionFactoryBuilder.getInstance().getSessionFactory().openSession()) {
            var transaction = session.beginTransaction();
            try {
                MutationQuery query = session.createMutationQuery(
                        "UPDATE PlayerData p SET p.lastOnline = :lastOnline WHERE p.id = :id");
                query.setParameter("lastOnline", playerData.getLastOnline());
                query.setParameter("id", playerData.getId());
                query.executeUpdate();
                transaction.commit();
            } catch (Exception e) {
                if (transaction.isActive()) {
                    transaction.rollback();
                }
            }
        }
    }
}
